drivers/onewire/onewire.c

```diff
@@ -0,0 +1,181 @@
+#define T_W_HOLD            (57)
+#define T_W_RECOVER         (10)
+#define T_R_SAMPLE          (7)
+#define T_R_RECOVER         (55)
+
+static inline int _read_bit(const onewire_t *owi)
+{
+    gpio_clear(owi->pin);
+    xtimer_usleep(T_RW_PULSE);
+    gpio_init(owi->pin, owi->imode);
+    xtimer_usleep(T_R_SAMPLE);
+    int in = (gpio_read(owi->pin)) ? 1 : 0;
+    xtimer_usleep(T_R_RECOVER);
+    gpio_init(owi->pin, owi->omode);
+    gpio_set(owi->pin);
+
+    return in;
+}
+
+static inline void _write_bit(const onewire_t *owi, int out)
+{
+    gpio_clear(owi->pin);
+    if (out) {
+        /* shift out a `1` */
+        xtimer_usleep(T_RW_PULSE);
+        gpio_set(owi->pin);
+        xtimer_usleep(T_W_HOLD);
+    }
+    else {
+        /* shift out a `0` */
+        xtimer_usleep(T_RW_PULSE + T_W_HOLD);
+        gpio_set(owi->pin);
+    }
+    xtimer_usleep(T_W_RECOVER);
+}
+
+int onewire_init(onewire_t *owi, const onewire_params_t *params)
+{
+    owi->pin = params->pin;
+
+    owi->omode = params->pin_mode;
+    owi->imode = (owi->omode == GPIO_OD_PU) ? GPIO_IN_PU : GPIO_IN;
+
+    if (gpio_init(owi->pin, owi->omode) != 0) {
+        return ONEWIRE_ERR_PINCFG;
+    }
+    gpio_set(owi->pin);
+    return ONEWIRE_OK;
+}
+
+int onewire_reset(const onewire_t *owi, const onewire_rom_t *rom)
+{
+    int res = ONEWIRE_OK;
+
+    gpio_clear(owi->pin);
+    xtimer_usleep(T_RESET_HOLD);
+    gpio_set(owi->pin);
+
+    xtimer_usleep(T_RESET_SAMPLE);
+    gpio_init(owi->pin, owi->imode);
+    if (gpio_read(owi->pin)) {
+        res = ONEWIRE_NODEV;
+    }
+
+    xtimer_usleep(T_RESET_HOLD - T_RESET_SAMPLE);
+    gpio_init(owi->pin, owi->omode);
+    gpio_set(owi->pin);
+
+    if ((res == ONEWIRE_OK) && (rom != NULL)) {
+        onewire_write_byte(owi, ONEWIRE_ROM_MATCH);
+        onewire_write(owi, rom->u8, sizeof(onewire_rom_t));
+    }
+
+    return res;
+}
+
+void onewire_read(const onewire_t *owi, void *data, size_t len)
+{
+    uint8_t *buf = (uint8_t *)data;
+
+    for (unsigned pos = 0; pos < (unsigned)len; pos++) {
+        /* read the next byte from the bus */
+        uint8_t tmp = 0;
+        for (int i = 0; i < 8; i++) {
+            tmp |= (_read_bit(owi) << i);
+        }
+        buf[pos] = tmp;
+    }
+}
+
+void onewire_write(const onewire_t *owi, const void *data, size_t len)
+{
+    uint8_t *buf = (uint8_t *)data;
+
+    for (unsigned pos = 0; pos < (unsigned)len; pos ++) {
+
+        for (int i = 0; i < 8; i++) {
+            _write_bit(owi, (buf[pos] & (1 << i)));
+        }
+    }
+}
+
+int onewire_search(const onewire_t *owi, onewire_rom_t *rom, int ld)
+{
+    /* initialize the search state */
+    if (ld == ONEWIRE_SEARCH_FIRST) {
+        memset(rom, 0, sizeof(onewire_rom_t));
+    }
+
+    /* send reset condition and issue search ROM command */
+    if (onewire_reset(owi, NULL) != ONEWIRE_OK) {
+        return ONEWIRE_NODEV;
+    }
+    onewire_write_byte(owi, ONEWIRE_ROM_SEARCH);
+
+    /* start search */
```

Comment:
Maybe add a short comment on how that search is performed.

Reply:
nothing to be explained in short, as it is quite complex. But will add a pointer to an explanation...